content/en/blog/_posts/2019-05-22-OPA-Gatekeeper-Policy-and-Governance-for-Kubernetes.md

```diff
@@ -0,0 +1,178 @@
+
+* Gatekeeper v1.0 - Uses OPA as the admission controller with the kube-mgmt sidecar enforcing configmap-based policies. It provides validating and mutating admission control. Donated by Styra. 
+* Gatekeeper v2.0 - Uses Kubernetes policy controller as the admission controller with OPA and kube-mgmt sidecars enforcing configmap-based policies. It provides validating and mutating admission control and audit functionality. Donated by Microsoft. 
+* Gatekeeper v3.0 - The admission controller is integrated with the [OPA constraint framework](https://github.com/open-policy-agent/frameworks/tree/master/constraint) to enforce CRD-based policies and allow declaratively configured policies to be reliably shareable. Built with kubebuilder, it provides validating and mutating (to be implemented) admission control and audit functionality. This enables the creation of policy templates for [Rego](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-write-policies/) policies, creation of policies as CRDs, and storage of audit results on policy CRDs. This project is a collaboration between Google, Microsoft, Red Hat, and Styra. 
+
+![](/images/blog/2019-05-22-opa-gatekeeper/v3.png)
+ 
+## Gatekeeper v3.0 Features 
+
+Now let’s take a closer look at the current state of Gatekeeper and how you can leverage all the latest features. Consider an organization that wants to ensure all objects in a cluster have departmental information provided as part of the object’s labels. How do we do this with Gatekeeper? 
+
+### Validating Admission Control 
+
+Once all the Gatekeeper components have been [installed](https://github.com/open-policy-agent/gatekeeper) in your cluster, the API server will trigger the Gatekeeper admission webhook to process the admission request whenever a resource in the cluster is created, updated, or deleted. 
+
+During the validation process, Gatekeeper will enforce all policies executed by OPA. Each policy is written with [Rego](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-write-policies/), a declarative query language used by OPA to enumerate instances of data that violate the expected state of the system. All policies are evaluated as a logical AND. If one policy is not satisfied, then the whole request is rejected. 
+
+### Policies and Constraints 
+
+With the integration of the OPA constraint framework, a policy is also known as a constraint. A constraint is a declaration that its author wants a system to meet a given set of requirements. Before defining a policy (aka constraint), you need to create a Policy Template (aka Constraint Template) that allows people to declare new policies (constraints). Each template describes both the [Rego](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-write-policies/) logic that enforces the policy and the schema for the policy, which includes the schema of the CRD and the parameters that can be passed into a policy, much like arguments to a function. 
+
```

Comment:
We should avoid conflating policy with constraint. Policy is an overloaded word in the community and carries no meaning on its own... Pod Security Policy, Scheduling Policy, Network Policy. When people mention policy enforcement out of context, I'm not sure anyone would know exactly what level of the stack we're talking about, or which set of teams should be interested.

Currently I'm thinking of "policy" as "this abstract state we want to enforce", such as "critical images are discoverable and securely built", and "constraints" as the concrete building blocks upon which that policy is defined "people must provide a `critical` label" AND "all pods where `critical == extremely` must pull from the `so.secure` repo"

Reply:
in short `policy (constraint)` -> `constraint`

and it may be useful to talk about the difference between policy and constraints, as that is a natural way to introduce the concepts and semantics.

Reply:
I updated this to use `constraint` where appropriate. Can you pls help draft 1 or 2 sentences to describe the difference policies vs constraints?

Reply:
@maxsmythe Can you pls help draft 1 or 2 sentences to describe the difference policies vs constraints?

Reply:
"a policy is also known as a constraint" is one example where the two words are considered equivalent, which is something that should be discouraged. To be clear, I'm not against the word "policy" showing up, but want to avoid the trap of under-explaining by using a word that has become too generic to be descriptive.